### ISSUE TITLE
Ceremony start time could delayed up to ~30 minutes

### ISSUE THREAD

**Issue by gurrpi** [quadratic-funding/mpc-phase2-suite#192]
Coordinator could specify desired ceremony starting time in seconds (e.g. 2022-11-13 01:12:28). But actually ceremony could started when 30 minutes periodic cronjob, which is responsible to start ceremony, has been triggered. 

For example,
* `startCeremony` Functions scheduled to run every 30 minutes from on time. It had been ran at 15:00. Then a coordinator set desired ceremony opening time at 15:01 and submitted. `startCeremony` would be triggered again at 15:30. It will open the ceremony and it had been delayed 29 minutes than coordinator's desired time.

**Comment by 0xjei** [comment]
The start and stop via scheduled Cloud Functions are designed to automate the start and stop tasks for the coordinator (i.e., otherwise the coordinator would have to issue a start/stop command manually on the opening/closing date and time of the ceremony).

The 30-minute schedule was not defined particularly strictly. Anyway, we need to find a way to avoid having too much start and stop useless calls.

**Comment by 0xjei** [comment]
UP: We could set the schedule to be 1 minute, forcing to a 1-minute max delay. That's what we can do in terms of automation. Thoughts on this? Or we should change the approach? :thinking:

**Comment by 0xjei** [comment]
> > UP: We could set the schedule to be 1 minute, forcing to a 1-minute max delay. That's what we can do in terms of automation. Thoughts on this? Or we should change the approach? thinking
> 
> There is requirement derived from current prompt to coordinator: _our service operates in punctual regarding user input (in seconds)._ In order to satisfy this requirement, we may need to either meet the expectation or modify it. Your idea appears to involve modifying the expectation of punctual operation to one that allows for a delay of up to 1 minute.

You are right, I like the way you explained the problem. I do not have a solution yet, just brainstorming below.

I'm assuming that we need a trigger to switch between ceremony states (open, close, finalized). We have different time-based information (startDate, endDate, etc.). 

- Current Approach: use a scheduled cloud function to open and close ceremonies. Max tollerance for delay is 1 minute.
- Possible Approach: The ceremony could be considered opened when the condition of `now > startDate && now < endDate` is met; we could switch from `ready` to `opened` state when the first user tries to join the waiting queue. The ceremony could be considered closed when the condition of `now > startDate && now > endDate` is met; we could switch from `opened` to `closed` when a user tries to join the waiting queue or the coordinator tries to finalize the ceremony.

Idk if you'd like the Possible Approach above, but it's the only alternative workflow that came up to my mind. Just because we need to trigger in some way this change, the alternative takes care of making the ceremony open/close without having a delay but relies on user/coordinator action (i.e., join the waiting queue or finalize).

**Comment by gurrpi** [comment]
> > > UP: We could set the schedule to be 1 minute, forcing to a 1-minute max delay. That's what we can do in terms of automation. Thoughts on this? Or we should change the approach? thinking
> > 
> > 
> > There is requirement derived from current prompt to coordinator: _our service operates in punctual regarding user input (in seconds)._ In order to satisfy this requirement, we may need to either meet the expectation or modify it. Your idea appears to involve modifying the expectation of punctual operation to one that allows for a delay of up to 1 minute.
> 
> You are right, I like the way you explained the problem. I do not have a solution yet, just brainstorming below.
> 
> I'm assuming that we need a trigger to switch between ceremony states (open, close, finalized). We have different time-based information (startDate, endDate, etc.).
> 
> * Current Approach: use a scheduled cloud function to open and close ceremonies. Max tollerance for delay is 1 minute.
> * Possible Approach: The ceremony could be considered opened when the condition of `now > startDate && now < endDate` is met; we could switch from `ready` to `opened` state when the first user tries to join the waiting queue. The ceremony could be considered closed when the condition of `now > startDate && now > endDate` is met; we could switch from `opened` to `closed` when a user tries to join the waiting queue or the coordinator tries to finalize the ceremony.
> 
> Idk if you'd like the Possible Approach above, but it's the only alternative workflow that came up to my mind. Just because we need to trigger in some way this change, the alternative takes care of making the ceremony open/close without having a delay but relies on user/coordinator action (i.e., join the waiting queue or finalize).

That is pretty neat, nice +1

**Comment by 0xjei** [comment]
> > > > UP: We could set the schedule to be 1 minute, forcing to a 1-minute max delay. That's what we can do in terms of automation. Thoughts on this? Or we should change the approach? thinking
> > > 
> > > 
> > > There is requirement derived from current prompt to coordinator: _our service operates in punctual regarding user input (in seconds)._ In order to satisfy this requirement, we may need to either meet the expectation or modify it. Your idea appears to involve modifying the expectation of punctual operation to one that allows for a delay of up to 1 minute.
> > 
> > 
> > You are right, I like the way you explained the problem. I do not have a solution yet, just brainstorming below.
> > I'm assuming that we need a trigger to switch between ceremony states (open, close, finalized). We have different time-based information (startDate, endDate, etc.).
> > 
> > * Current Approach: use a scheduled cloud function to open and close ceremonies. Max tollerance for delay is 1 minute.
> > * Possible Approach: The ceremony could be considered opened when the condition of `now > startDate && now < endDate` is met; we could switch from `ready` to `opened` state when the first user tries to join the waiting queue. The ceremony could be considered closed when the condition of `now > startDate && now > endDate` is met; we could switch from `opened` to `closed` when a user tries to join the waiting queue or the coordinator tries to finalize the ceremony.
> > 
> > Idk if you'd like the Possible Approach above, but it's the only alternative workflow that came up to my mind. Just because we need to trigger in some way this change, the alternative takes care of making the ceremony open/close without having a delay but relies on user/coordinator action (i.e., join the waiting queue or finalize).
> 
> That is pretty neat, nice +1

Cool, glad you liked it. I just need to better explore some possible corner cases and then validate the idea before starting to work on it.
